substrate/in-instructions/primitives/src/lib.rs

```diff
@@ -0,0 +1,47 @@
+#![cfg_attr(docsrs, feature(doc_cfg))]
+#![cfg_attr(docsrs, feature(doc_auto_cfg))]
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use scale::{Encode, Decode, MaxEncodedLen};
+use scale_info::TypeInfo;
+
+#[cfg(feature = "std")]
+use serde::{Serialize, Deserialize};
+
+use sp_core::{ConstU32, bounded::BoundedVec};
+
+// Monero, our current longest address candidate, has a longest address of featured with payment ID
+// 1 (enum) + 1 (flags) + 64 (two keys) + 8 (payment ID) = 74
+pub const MAX_ADDRESS_LEN: u32 = 74;
+// Should be enough for a Uniswap v3 call
+pub const MAX_DATA_LEN: u32 = 512;
+
+#[derive(Clone, PartialEq, Eq, Debug, Encode, Decode, MaxEncodedLen, TypeInfo)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+pub struct ExternalAddress(BoundedVec<u8, ConstU32<{ MAX_ADDRESS_LEN }>>);
+impl ExternalAddress {
+  #[cfg(feature = "std")]
+  pub fn new(address: Vec<u8>) -> Result<ExternalAddress, &'static str> {
+    Ok(ExternalAddress(address.try_into().map_err(|_| "address length exceeds {MAX_ADDRESS_LEN}")?))
+  }
+
+  pub fn address(&self) -> &[u8] {
+    self.0.as_ref()
+  }
+
+  #[cfg(feature = "std")]
+  pub fn consume(self) -> Vec<u8> {
+    self.0.into_inner()
+  }
+}
+
+// Not "in" as "in" is a keyword
+mod incoming;
```

Comment:
txin

Reply:
I don't necessarily hate it? It's just awkward since we don't handle transactions in. We handle outputs/coins/messages in. That's why this was vague to "incoming" (or just "in", if that wasn't a key word).

We can edit to txin to be closer to the BTC standard, yet that raises its own discussions on the BTC standard being garbage.

Do you have any reason to dislike "incoming" so much or yet another option available?

Reply:
Incoming is too ambiguous. I'm fine with OutputsIn or CoinsIn, just something other than incoming.

Reply:
Except these aren't outputs and they're not coins. They're instructions. It's just annoying as hell to write `in_instructions::primitives::in_instructions::InInstruction` versus `in_instructions::primitives::in::InInstruction`. The members within, and the crate name, provide sufficient context IMO.

Also, this discussion is moot. This is literally a file name. It's written as `mod incoming; pub use incoming::*;` Accordingly, our public API is just `in_instructions::primitives::{InInstruction, OutInstruction}`. "incoming" is never in the public API. Can we drop this accordingly, please?